File: core/src/passes/licm.cpp
// core/src/passes/licm.cpp — F2-PS-003 Loop-Invariant Code Motion.
//
// For each natural loop in the function, find statements whose result
// does not change across iterations and hoist them to the loop's
// pre-header. A statement qualifies as loop-invariant iff:
//
//   1. It is pure (no observable side effect — the same set DCE uses).
//   2. All of its operand Refs are defined outside the loop body.
//
// Iteration: we re-scan each loop's body after a successful hoist,
// because the just-hoisted stmt's result is now "outside the body" and
// may unlock further hoists. This converges in O(stmts_in_loop) per
// loop (each stmt can leave the body at most once).
//
// Pre-header detection is conservative: the loop is hoisted iff the
// header has exactly one CFG predecessor outside the loop body. If
// the natural-loop entry has multiple incoming non-loop edges (rare
// in compiler output but possible in hand-written assembly), we skip
// — synthesising a new preheader block changes CFG shape and is
// deferred to a follow-up.
//
// **Wiring caveat (carries from F2-PS-004 / Global CSE):** today's
// translator emits single-block functions, which trivially have zero
// loops. LICM on those is a no-op. The plumbing here is the
// deliverable; the algorithmic gain unlocks once
// `core/src/translator/` gains multi-instruction fusion or the
// translator learns to merge a backward-jump target with its
// predecessor.

#include "prisma/passes.hpp"

#include <algorithm>
#include <unordered_set>
#include <variant>

#include "prisma/dominators.hpp"

namespace prisma::passes {

namespace {

bool stmt_is_pure(const ir::Op& op) noexcept {
    return std::visit([](auto const& x) -> bool {
        using T = std::decay_t<decltype(x)>;
        return std::is_same_v<T, ir::Constant>
            || std::is_same_v<T, ir::LoadReg>
            || std::is_same_v<T, ir::LoadSegBase>
            || std::is_same_v<T, ir::BinOp>
            || std::is_same_v<T, ir::Compare>
            || std::is_same_v<T, ir::Extend>
            || std::is_same_v<T, ir::Truncate>;
    }, op);
}

void collect_operand_refs(const ir::Op& op, std::vector<ir::Ref>& into) {
    std::visit([&](auto const& x) {
        using T = std::decay_t<decltype(x)>;
        if constexpr (std::is_same_v<T, ir::BinOp>) {
            into.push_back(x.lhs); into.push_back(x.rhs);
        } else if constexpr (std::is_same_v<T, ir::Compare>) {
            into.push_back(x.lhs); into.push_back(x.rhs);
        } else if constexpr (std::is_same_v<T, ir::Extend>) {
            into.push_back(x.value);
        } else if constexpr (std::is_same_v<T, ir::Truncate>) {
            into.push_back(x.value);
        }
        // Constant / LoadReg / LoadSegBase have no operand refs.
    }, op);
}

}  // namespace

ir::Function loop_invariant_motion(const ir::Function& fn) {
    if (fn.blocks.empty()) return fn;

    const auto loops = ir::natural_loops(fn);
    if (loops.empty()) return fn;

    // id → position index, used to address `out.blocks` directly.
    std::unordered_map<std::uint32_t, std::size_t> idx;
    idx.reserve(fn.blocks.size());
    for (std::size_t i = 0; i < fn.blocks.size(); ++i) {
        idx[fn.blocks[i].id] = i;
    }

    // Predecessor map for preheader detection.
    std::unordered_map<std::uint32_t, std::vector<std::uint32_t>> preds;
    for (const auto& b : fn.blocks) {
        for (auto s : ir::successors(fn, b.id)) preds[s].push_back(b.id);
    }

    ir::Function out = fn;

    for (const auto& loop : loops) {
        // Set of block ids in this loop, for fast membership checks.
        std::unordered_set<std::uint32_t> body_ids(
            loop.body.begin(), loop.body.end());

        // Identify the preheader: the unique non-loop predecessor of
        // `loop.header`. Skip the whole loop if 0 or >1 candidates.
        const auto pit = preds.find(loop.header);
        if (pit == preds.end()) continue;
        std::vector<std::uint32_t> outside_preds;
        for (auto p : pit->second) {
            if (body_ids.find(p) == body_ids.end()) {
                outside_preds.push_back(p);
            }
        }
        if (outside_preds.size() != 1) continue;
        const std::uint32_t preheader_id = outside_preds.front();
        const auto preh_it = idx.find(preheader_id);
        if (preh_it == idx.end()) continue;
        auto& preheader = out.blocks[preh_it->second];
        if (preheader.stmts.empty()) continue;  // no terminator slot

        // Set of Refs defined in the loop body. Rebuilt each pass
        // because successful hoists shrink it.
        auto rebuild_body_refs = [&](std::unordered_set<ir::Ref>& set) {
            set.clear();
            for (auto bid : loop.body) {
                const auto i = idx[bid];
                for (const auto& st : out.blocks[i].stmts) {
                    if (st.result) set.insert(*st.result);
                }
            }
        };

        std::unordered_set<ir::Ref> body_refs;
        rebuild_body_refs(body_refs);

        // Iterate to fixed point per loop.
        bool changed = true;
        while (changed) {
            changed = false;
            for (auto bid : loop.body) {
                auto& block = out.blocks[idx[bid]];
                // Don't hoist the terminator (last stmt of any block).
                if (block.stmts.empty()) continue;
                const std::size_t terminator_pos = block.stmts.size() - 1;
                std::size_t i = 0;
                while (i < terminator_pos) {
                    const auto& stmt = block.stmts[i];
                    if (!stmt.result || !stmt_is_pure(stmt.op)) {
                        ++i;
                        continue;
                    }
                    std::vector<ir::Ref> operand_refs;
                    collect_operand_refs(stmt.op, operand_refs);
                    bool all_outside = true;
                    for (auto r : operand_refs) {
                        if (body_refs.count(r) != 0) {
                            all_outside = false;
                            break;
                        }
                    }
                    if (!all_outside) {
                        ++i;
                        continue;
                    }
                    // Hoist: erase from body block, insert at the
                    // preheader's end-1 (before its terminator). Save
                    // the result Ref before the move-out so we can
                    // update body_refs without touching the moved-from
                    // statement.
                    const ir::Ref hoisted_result = *stmt.result;
                    ir::Stmt hoisted = std::move(block.stmts[i]);
                    block.stmts.erase(block.stmts.begin()
                                      + static_cast<std::ptrdiff_t>(i));
                    auto& ph_stmts = preheader.stmts;
                    if (ph_stmts.empty()) {
                        ph_stmts.push_back(std::move(hoisted));
                    } else {
                        ph_stmts.insert(
                            ph_stmts.begin()
                                + static_cast<std::ptrdiff_t>(
                                    ph_stmts.size() - 1),
                            std::move(hoisted));
                    }
                    body_refs.erase(hoisted_result);
                    changed = true;
                    // Do not increment i; the next stmt has shifted up.
                }
            }
        }
    }

    return out;
}

}  // namespace prisma::passes

File: shell/prisma-passes/src/licm.rs
//! Loop-invariant code motion.
//!
//! Mirrors C++ `loop_invariant_motion`. For each natural loop, hoist pure
//! statements whose operands are all defined outside the loop body into the
//! loop's preheader (the unique non-loop predecessor of the header). Re-scans
//! to a fixed point per loop, since a hoist can unlock further hoists.
//!
//! Conservative: skips a loop unless its header has exactly one outside
//! predecessor (no synthetic preheader insertion). Terminators are never
//! hoisted. Today's translator emits single-block functions (zero loops), so
//! this is plumbing — the gain unlocks once multi-block loops are emitted.

use std::collections::{HashMap, HashSet};

use prisma_ir::{Function, Op, Ref};

use crate::cfg;
use crate::Pass;

/// Purity set for hoisting (matches C++ `stmt_is_pure` in licm.cpp — narrower
/// than DCE's: only side-effect-free, position-independent value computations).
fn stmt_is_pure(op: &Op) -> bool {
    matches!(
        op,
        Op::Constant(_)
            | Op::LoadReg(_)
            | Op::LoadSegBase(_)
            | Op::BinOp(_)
            | Op::Compare(_)
            | Op::Extend(_)
            | Op::Truncate(_)
    )
}

fn operand_refs(op: &Op, into: &mut Vec<Ref>) {
    match op {
        Op::BinOp(x) => {
            into.push(x.lhs);
            into.push(x.rhs);
        }
        Op::Compare(x) => {
            into.push(x.lhs);
            into.push(x.rhs);
        }
        Op::Extend(x) => into.push(x.value),
        Op::Truncate(x) => into.push(x.value),
        // Constant / LoadReg / LoadSegBase have no operand refs.
        _ => {}
    }
}

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
pub struct Licm;

impl Licm {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl Pass for Licm {
    fn name(&self) -> &'static str {
        "loop_invariant_motion"
    }

    fn run(&self, func: Function) -> Function {
        loop_invariant_motion(func)
    }
}

/// Hoist loop-invariant statements to loop preheaders.
pub fn loop_invariant_motion(mut func: Function) -> Function {
    if func.blocks.is_empty() {
        return func;
    }
    let loops = cfg::natural_loops(&func);
    if loops.is_empty() {
        return func;
    }

    let idx: HashMap<u32, usize> = func
        .blocks
        .iter()
        .enumerate()
        .map(|(i, b)| (b.id, i))
        .collect();

    // Predecessor map for preheader detection.
    let mut preds: HashMap<u32, Vec<u32>> = HashMap::new();
    for b in &func.blocks {
        for s in cfg::successors(&func, b.id) {
            preds.entry(s).or_default().push(b.id);
        }
    }

    for loop_ in &loops {
        let body_ids: HashSet<u32> = loop_.body.iter().copied().collect();

        // Preheader = unique non-loop predecessor of the header.
        let Some(hpreds) = preds.get(&loop_.header) else {
            continue;
        };
        let outside: Vec<u32> = hpreds
            .iter()
            .copied()
            .filter(|p| !body_ids.contains(p))
            .collect();
        if outside.len() != 1 {
            continue;
        }
        let preheader_id = outside[0];
        let Some(&preheader_pos) = idx.get(&preheader_id) else {
            continue;
        };
        if func.blocks[preheader_pos].stmts.is_empty() {
            continue; // no terminator slot
        }

        // Refs defined in the loop body; shrinks as we hoist.
        let mut body_refs: HashSet<Ref> = HashSet::new();
        for &bid in &loop_.body {
            for st in &func.blocks[idx[&bid]].stmts {
                if let Some(r) = st.result {
                    body_refs.insert(r);
                }
            }
        }

        let mut changed = true;
        while changed {
            changed = false;
            for &bid in &loop_.body {
                let bpos = idx[&bid];
                loop {
                    let term_pos = {
                        let n = func.blocks[bpos].stmts.len();
                        if n == 0 {
                            break;
                        }
                        n - 1
                    };
                    // Find the first hoistable statement before the terminator.
                    let mut hoist_at: Option<usize> = None;
                    for i in 0..term_pos {
                        let stmt = &func.blocks[bpos].stmts[i];
                        if stmt.result.is_none() || !stmt_is_pure(&stmt.op) {
                            continue;
                        }
                        let mut ops = Vec::new();
                        operand_refs(&stmt.op, &mut ops);
                        if ops.iter().all(|r| !body_refs.contains(r)) {
                            hoist_at = Some(i);
                            break;
                        }
                    }
                    let Some(i) = hoist_at else { break };

                    // Remove from body, insert before the preheader terminator.
                    let hoisted = func.blocks[bpos].stmts.remove(i);
                    let hoisted_result = hoisted.result;
                    let ph = &mut func.blocks[preheader_pos].stmts;
                    let insert_pos = ph.len() - 1;
                    ph.insert(insert_pos, hoisted);
                    if let Some(r) = hoisted_result {
                        body_refs.remove(&r);
                    }
                    changed = true;
                }
            }
        }
    }

    func
}

#[cfg(test)]
mod tests {
    use super::*;
    use prisma_ir::{BasicBlock, BinOp, BinOpKind, CondJump, Constant, Jump, OpSize, Return, Stmt};

    #[test]
    fn no_loops_is_noop() {
        let f = Function {
            entry: 0,
            blocks: vec![BasicBlock {
                id: 0,
                stmts: vec![Stmt::new(None, Op::Return(Return))],
            }],
        };
        assert_eq!(loop_invariant_motion(f.clone()), f);
    }

    #[test]
    fn hoists_invariant_binop_to_preheader() {
        // block0 (preheader): r0=const; r1=const; jump 1
        // block1 (loop header): r2 = add r0,r1 (invariant); cond 1/2
        // block2: ret
        // r2 should hoist into block0 before its terminator.
        let f = Function {
            entry: 0,
            blocks: vec![
                BasicBlock {
                    id: 0,
                    stmts: vec![
                        Stmt::new(
                            Some(0),
                            Op::Constant(Constant {
                                value: 3,
                                size: OpSize::I64,
                            }),
                        ),
                        Stmt::new(
                            Some(1),
                            Op::Constant(Constant {
                                value: 4,
                                size: OpSize::I64,
                            }),
                        ),
                        Stmt::new(None, Op::Jump(Jump { target_block: 1 })),
                    ],
                },
                BasicBlock {
                    id: 1,
                    stmts: vec![
                        Stmt::new(
                            Some(2),
                            Op::BinOp(BinOp {
                                op: BinOpKind::Add,
                                lhs: 0,
                                rhs: 1,
                                size: OpSize::I64,
                            }),
                        ),
                        Stmt::new(
                            None,
                            Op::CondJump(CondJump {
                                cond: 2,
                                if_true: 1,
                                if_false: 2,
                            }),
                        ),
                    ],
                },
                BasicBlock {
                    id: 2,
                    stmts: vec![Stmt::new(None, Op::Return(Return))],
                },
            ],
        };
        let out = loop_invariant_motion(f);
        // block0 now holds the hoisted add before its Jump terminator.
        let ph = &out.blocks[0].stmts;
        assert_eq!(ph.len(), 4);
        assert!(
            matches!(ph[2].op, Op::BinOp(_)),
            "add hoisted into preheader"
        );
        assert!(matches!(ph[3].op, Op::Jump(_)), "terminator stays last");
        // The loop header no longer contains the add (only its terminator).
        assert_eq!(out.blocks[1].stmts.len(), 1);
        assert!(matches!(out.blocks[1].stmts[0].op, Op::CondJump(_)));
    }

    #[test]
    fn variant_binop_is_not_hoisted() {
        // r2 depends on r2 (loop-carried via self) — not invariant.
        let f = Function {
            entry: 0,
            blocks: vec![
                BasicBlock {
                    id: 0,
                    stmts: vec![Stmt::new(None, Op::Jump(Jump { target_block: 1 }))],
                },
                BasicBlock {
                    id: 1,
                    stmts: vec![
                        Stmt::new(
                            Some(2),
                            Op::BinOp(BinOp {
                                op: BinOpKind::Add,
                                lhs: 2,
                                rhs: 2,
                                size: OpSize::I64,
                            }),
                        ),
                        Stmt::new(
                            None,
                            Op::CondJump(CondJump {
                                cond: 2,
                                if_true: 1,
                                if_false: 2,
                            }),
                        ),
                    ],
                },
                BasicBlock {
                    id: 2,
                    stmts: vec![Stmt::new(None, Op::Return(Return))],
                },
            ],
        };
        let out = loop_invariant_motion(f.clone());
        assert_eq!(out, f, "loop-carried value must not be hoisted");
    }
}
